incident/mysql/incident_test.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		t.Fatalf("failed to query rows: %v", err)
+	}
+	defer rows.Close()
+
+	var got []entry
+	for rows.Next() {
+		var e entry
+		if err := rows.Scan(&e.BaseURL, &e.Summary, &e.Category, &e.FullURL, &e.Details); err != nil {
+			t.Fatalf("failed to scan row: %v", err)
+		}
+		got = append(got, e)
+	}
+	if err := rows.Err(); err != nil {
+		t.Errorf("incident table iteration failed: %v", err)
+	}
+	if diff := cmp.Diff(got, want); diff != "" {
+		t.Errorf("incident table: diff (-got +want)\n%s", diff)
+	}
+}
+
+func TestLogf(t *testing.T) {
+	ctx := context.Background()
+	cleanTestDB(ctx)
+
+	checkContents(ctx, t, nil)
+
+	reporter, err := NewMySQLReporter(ctx, testDB, "unittest")
+	if err != nil {
+		t.Fatalf("failed to build MySQLReporter: %v", err)
+	}
+	e := entry{BaseURL: "base", Summary: "summary", Category: "signature", FullURL: "full", Details: "blah"}
+
+	reporter.Log(ctx, e.BaseURL, e.Summary, e.Category, e.FullURL, e.Details)
+	checkContents(ctx, t, []entry{e})
+
+	reporter.Log(ctx, e.BaseURL, e.Summary, e.Category, e.FullURL, e.Details)
+	checkContents(ctx, t, []entry{e, e})
+
+	reporter.Logf(ctx, e.BaseURL, e.Summary, e.Category, e.FullURL, "%s", e.Details)
+	checkContents(ctx, t, []entry{e, e, e})
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if err := mySQLAvailable(); err != nil {
+		glog.Errorf("MySQL not available, skipping all MySQL storage tests: %v", err)
+		return
+	}
+	ctx := context.Background()
+	var err error
+	testDB, err = newIncidentDB(ctx)
+	if err != nil {
+		glog.Exitf("failed to create test database: %v", err)
+	}
+	defer testDB.Close()
+	cleanTestDB(ctx)
+	ec := m.Run()
+	os.Exit(ec)
+}
+
+var (
+	testDB      *sql.DB
+	dataSource  = "root@/"
+	incidentSQL = "incident.sql"
+)
+
+// mySQLAvailable indicates whether a default MySQL database is available.
+func mySQLAvailable() error {
+	db, err := sql.Open("mysql", dataSource)
+	if err != nil {
+		return fmt.Errorf("sql.Open(): %v", err)
+	}
+	defer db.Close()
+	if err := db.Ping(); err != nil {
+		return fmt.Errorf("db.Ping(): %v", err)
+	}
+	return nil
+}
+
+// newEmptyDB creates a new, empty database.
+func newEmptyDB(ctx context.Context) (*sql.DB, error) {
+	db, err := sql.Open("mysql", dataSource)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a randomly-named database and then connect using the new name.
+	name := fmt.Sprintf("mono_%v", time.Now().UnixNano())
+
+	stmt := fmt.Sprintf("CREATE DATABASE %v", name)
+	if _, err := db.ExecContext(ctx, stmt); err != nil {
+		return nil, fmt.Errorf("error running statement %q: %v", stmt, err)
+	}
+	db.Close()
+
+	db, err = sql.Open("mysql", dataSource+name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open new database %q: %v", name, err)
+	}
+	return db, db.Ping()
+}
+
+// newIncidentDB creates an empty database with the incident schema.
+func newIncidentDB(ctx context.Context) (*sql.DB, error) {
```

Comment:
In the Trillian repo, we instead assume the presence of a test database with the correct schema and use that, right? What's the argument for doing it differently here?

Reply:
This is consistent with (at least some of)) the Trillian repo: https://github.com/google/trillian/blob/master/storage/testdb/testdb.go#L82

(and this approach seemed more hermetic)